### PR TITLE
Allow composite to be a sub-aggregation of filter

### PR DESCRIPTION
**Pull request by not-napoleon**
This PR is intended to replace #38863.  After this change, composite can be an arbitrarily deep child aggregation of any combination of nested and filter aggregations.

Relates to #37178

### CODE REVIEW

**Review by nik9000 — COMMENTED** [review]
Could you add something to `CompositeAggregatorTests`? Its nice to have this there too so its easier to use a debugger and stuff.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/230_composite.yml [hunk 1]
```diff
@@ -1120,3 +1120,44 @@ setup:
...
   - match: { aggregations.courses.sessions.names.buckets.1.doc_count: 4}
+
+---
+"Nested then filter then nested then terms":
+  - do:
+      search:
+        rest_total_hits_as_int: true
+        index: verynested
+        body:
+          "aggregations": {
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/230_composite.yml:1131]: I can change it if you feel strongly.  I was copying back and forth between dev console while tinkering with the tests, that's why it's in json.

**Comment by nik9000** [comment]
The rest compat failure just need master merged in.

I did talk to @not-napoleon about adding a test for when the parent is just a `filter` agg. This is probably a silly thing to do - you can just put the query on top level query instead of the filter agg. But the code he's written allows it. And, surprise! It spits out the wrong results.

**Comment by not-napoleon** [comment]
yeah, good call on testing that.  I'm working on a fix for it now.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Review by nik9000 — APPROVED** [review]
Makes sense to me.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Comment by BountyCountry** [comment]
@not-napoleon I don't see this one in the 7.13.0 release notes. Did it make it?

**Comment by not-napoleon** [comment]
It should be there.  I see it in the git log for 7.13, and we would have updated the tags on this issue if we'd bumped it forward.  Not sure why it didn't make the release notes though.

**Pull request merged**